File: sdk/elixir/lib/bsdkrun/images.ex
defmodule Bsdkrun.Images do
  @moduledoc "Image operations: list downloaded OCI + fetched BSD images."

  alias Bsdkrun.{Cli, Error, Types}
  alias Bsdkrun.Types.ImageInfo

  @doc "List downloaded images."
  @spec list() :: {:ok, [ImageInfo.t()]} | {:error, Error.t()}
  def list do
    with {:ok, res} <- Cli.checked(["images", "--json"], "bsdkrun images") do
      rows =
        res.stdout
        |> then(&if(String.trim(&1) == "", do: "[]", else: &1))
        |> Jason.decode!()
        |> Enum.map(&Types.image_info/1)

      {:ok, rows}
    end
  end
end

File: sdk/clojure/src/bsdkrun/images.clj
(ns bsdkrun.images
  "Host-level image operations. Mirrors `sdk/ruby/lib/bsdkrun/images.rb`."
  (:refer-clojure :exclude [list])
  (:require [bsdkrun.process :as process]
            [bsdkrun.types :as types]))

(defn list
  "List downloaded images (pulled OCI images + fetched BSD images)."
  []
  (let [res (process/run! ["images" "--json"] {:label "bsdkrun images"})]
    (mapv types/image-info-from-row (types/read-json-rows (:stdout res)))))
